src/OpenTelemetry.Instrumentation.Runtime/description.md

```diff
@@ -0,0 +1,198 @@
+| process.memory.usage    | The amount of physical memory in use   | `By`           | ObservableGauge   | `Int64`    |                  |                  |
+| process.memory.virtual  | The amount of committed virtual memory | `By`           | ObservableGauge   | `Int64`    |                  |                  |
+
+- CPU utilization
+  - [Process.TotalProcessorTime](https://docs.microsoft.com/dotnet/api/system.diagnostics.process.totalprocessortime)
+  divided by ([Environment.ProcessorCount](https://docs.microsoft.com/dotnet/api/system.environment.processorcount)
+  \* ([DateTime.Now](https://docs.microsoft.com/dotnet/api/system.datetime.now) -
+  [Process.StartTime](https://docs.microsoft.com/dotnet/api/system.diagnostics.process.starttime)))
+
+- CPU Time:
+  - [Process.UserProcessorTime](https://docs.microsoft.com/dotnet/api/system.diagnostics.process.userprocessortime):
+  Gets the user processor time for this process.
+  - [Process.PrivilegedProcessorTime](https://docs.microsoft.com/dotnet/api/system.diagnostics.process.privilegedprocessortime):
+  Gets the privileged processor time for this process.
+
+- [Environment.ProcessorCount](https://docs.microsoft.com/dotnet/api/system.environment.processorcount):
+  Gets the number of processors available to the current process.
+- Memory usage: [Process.GetCurrentProcess().WorkingSet64](https://docs.microsoft.com/dotnet/api/system.diagnostics.process.workingset64):
+  Gets the amount of physical memory, in bytes, allocated for the currently
+  active process.
+- Memory virtual: [Process.GetCurrentProcess().VirtualMemorySize64](https://docs.microsoft.com/dotnet/api/system.diagnostics.process.virtualmemorysize64):
+  Gets the amount of the virtual memory, in bytes, allocated for the currently
+  active process.
+
+Question: EventCounter implementation exposes a metric named `working-set` with
+`Environment.WorkingSet`. Is it equal to `Process.GetCurrentProcess().WorkingSet64`
+property? I need to decide on which is more suitable for showing users the memory
+usage for the process, or whether to include both.
+
+- [Environment.WorkingSet](https://docs.microsoft.com/en-us/dotnet/api/system.environment.workingset?view=net-6.0):
+  A 64-bit signed integer containing the number of bytes of physical memory mapped
+  to the process context.
+
+## Assemblies related metrics
+
+The metrics in this section can be enabled by setting the
+`RuntimeMetricsOptions.IsAssembliesEnabled` switch.
+
+| Name                                      | Description                 | Units          | Instrument Type   | Value Type | Attribute Key(s) | Attribute Values |
+|-------------------------------------------|-----------------------------|----------------|-------------------|------------|------------------|------------------|
+| process.runtime.dotnet.**assembly.count** | Number of Assemblies Loaded | `{assemblies}` | ObservableCounter | `Int64`    |                  |                  |
```

Comment:
.NET runtime has the ability to [unload assemblies](https://docs.microsoft.com/en-us/dotnet/standard/assembly/load-unload), so `AppDomain.GetAssemblies().Length` could decrease in some occasions. The instrument type should be `ObservableUpDownCounter` (before `ObservableUpDownCounter` is ready, I'll use `ObservableGauge` in implementation). I'll update the doc later.

Reply:
Your comment is for .NET Framework and not for .NET Core as .NET Core doesn't allow more than one appomain in the process. 

> AppDomain.GetAssemblies().Length could decrease in some occasions.

I am not sure if this is accurate either. You may unload the whole appdomain and not individual assembly. So, technically the `AppDomain.GetAssemblies().Length` doesn't decrease. but the total assemblies in the process can decrease. 

In .NET Core, uses AssemblyLoadContext for unloading all assemblies in that context by unloading the context. 

CC @noahfalk

Reply:
I haven't tested it, but I would assume that assemblies in any AssemblyLoadContext, including an unloadable one, appear in the AppDomain.GetAssemblies() list. In that case AppDomain.GetAssemblies().Length could decrease because a load context unloaded.

Reply:
Will such a counter have really a useful meaning in .NET core apps? And how can users handle such a counter in .NET Framework apps? or we don't care much about .NET Framework cases?

Reply:
My answer is YES. I can think of a valid use case - certain online service / background job run out of memory after hours and the engineers can tell from the "assembly loaded" metrics that the process kept loading plugins without unloading them.

This happened in my previous team, where a backend processing job detects new XML schema and compiles it to assembly (for sake of perf) and load it dynamically, the backend job ended up loading tens of thousands of assemblies.

Reply:
Sounds good then!

Reply:
Thanks, Tarek and Noah! I'm getting the assembly count for the "current domain" `AppDomain.CurrentDomain.GetAssemblies().Length`, and the metric has to work for both .NET Framework and .NET Core. If the value could decrease in any case, I couldn't use `ObservableCounter` because a metric data point [could be dropped](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/383).

I found one [remark](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain?view=net-6.0#remarks) about a special case when the assembly is in the default application domain:
> If an assembly is loaded into the default application domain, it cannot be unloaded from memory while the process is running.

If Noah's right about assemblies in AssemblyLoadContext would appear in `AppDomain.CurrentDomain.GetAssemblies()`, then I should use `ObservableGauge` (and use `ObservableUpDownCounter` when available). Otherwise it might lead to loss of data.

Reply:
> I'm getting the assembly count for the "current domain" AppDomain.CurrentDomain.GetAssemblies().Length

If we are reporting the number from the current appdomain only, I'll question the @reyang's scenario `certain online service / background job run out of memory after hours and the engineers can tell from the "assembly loaded" metrics that the process kept loading plugins without unloading them.` when running on .NET Framework. Looks what we report wouldn't be enough for that scenario. 

To clarify, .NET Framework can have multiple appdomains. I guess it will be more useful reporting the total count of assemblies in all appdomain and not necessary the current appdomain. 

In .NET Core, there is only one appdomain, so reporting the number from the current appdomain would work, I guess. We need to confirm loading and unloading AssemblyLoadContext is going to be reflected to the number we report. When I get some time, I can try to do that.

Reply:
Updated the instrument type to `ObservableGauge`.

For implementation on getting all AppDomains, I found three articles [1](http://miteshsureja.blogspot.com/2012/02/hot-to-get-list-of-all-appdomains.html), [2](https://stackoverflow.com/questions/33890264/enumerate-all-appdomains-without-mscoree) and [3](CorRuntimeHost), but they all seem like very old school code depending on MSCorEE.dll.

I still need to figure out whether I could use those APIs, or find better alternatives.